Multi-namespace monitors

davekonopka opened the pull request:
This adds support for creating multiple monitors from a single definition for a list of namespaces.

The underlying Rodd monitor must have `vary_by_namespace: true` set. The monitor's namespace definition value can be a string or a list. Each namespace definition value will produce a separate monitor and Terraform file.

I had to reorganize the Rodd classes to get this working. Now the Pentagon component builds up a collection of monitors, applies overrides in sequential order, and then writes out the results to TF files.

Resolves #37

davekonopka commented:
@ejether @mjhuber This PR is fully working now and ready for review.

mjhuber commented:
Hi Dave, thanks for contributing!  This will be a huge timesaver to be able to create multiple monitors for different namespaces.

I had some trouble getting the files to generate when I was testing.  Given this yml file:
```
definitions:
  environment: prod
  notifications: '@testtesttest'
  namespace: kube-system
dashboards:
  - source: kubernetes.resources
monitors:
  - source: kubernetes
downtimes: []
exceptions:
  - kubernetes.cluster_bytes_received
```

it generated prod_kubelet_unhealthy, which I think was the last definition in the list.
[output.log](https://github.com/reactiveops/rodd/files/2956928/output.log)

davekonopka commented:
@mjhuber That's odd. I'm getting a different result for the same config yaml. The following monitor tf's are generated.

This PR is a different branch than my earlier one. Is it possible you're on the old branch?

```Untracked files:
  (use "git add <file>..." to include in what will be committed)
	prod_clock_in_sync_with_ntp.tf
	prod_deployment_replica_alert_kube_system.tf
	prod_high_system_load_avg_on_cluster.tf
	prod_host_disk_usage_high.tf
	prod_i_o_wait_times_are_high.tf
	prod_increase_in_network_errors.tf
	prod_increased_kube_system_crashes.tf
	prod_kubelet_unhealthy.tf
	prod_node_is_not_ready.tf
	prod_pods_are_stuck_pending.tf
	prod_running_out_of_memory.tf```

mjhuber commented:
Yep, i'm using this branch.  are you using `pentagon add datadog.montiors -f <filename>` to generate them?

davekonopka commented:
Yup, that's the command I'm running. I'll take a look and see if I can reproduce what you're seeing.

davekonopka commented:
@mjhuber I was able to reproduce the same behavior by loading the earlier PR branch. Once I ran `pip install -e .` on the new branch I'm not having the problem. Maybe we could take a look together?

davekonopka commented:
I started a fresh virtualenv, ran `pip install -e .`, and `pentagon add datadog.monitors -f rodd.yml` and it worked as expected.

mjhuber commented:
so strange.  I created a new virtualenv and ran the steps below, still having the problem.

```
virtualenv -p /usr/local/bin/python new-rodd
cd new-rodd/ && source bin/activate
pip install git+https://github.com/reactiveops/pentagon.git
pip install git+ssh://git@github.com/reactiveops/pentagon-workstation-ng.git
pip install git+ssh://git@github.com/reactiveops/rodd.git@dk/namespaced-monitors
```

```
huberm@mbp01:~/code/venvs/new-rodd$ pip freeze | grep pentagon-datadog
DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
pentagon-datadog==1.11.0
```

davekonopka commented:
Following those steps reproduced it for me. I'll dig in further.

davekonopka commented:
@mjhuber I tracked down the issue. I was using the source path to generate a unique key during pre-processing. Under virtualenv `python2.7` in the path was resulting in duplicate keys for all the monitors. I simplified the key to avoid periods being an issue and it's fixed now.

I also adjusted the kube-system pod crash monitor to support multiple namespaces too.
